Feat/merge trains additional #2547

isaac-philip opened the pull request:
## Merge Trains Api additional support added :

1. Status of Merge Request on Merge Train
2. Add Merge Request to Merge Train

closes #2547 

_(can work to update the MR with docs till its reviewed/ required, thanks)_

isaac-philip committed:
feat(functional): merge train -> merge request :> status fetch

isaac-philip committed:
feat(functional): merge train -> merge request :> status fetch

isaac-philip committed:
feat(functional): #2547 merge-train api for status and add mr

Signed-off-by: Isaac Philip <4974658+isaac-philip@users.noreply.github.com>

isaac-philip committed:
feat(functional): #2547 merge-train api for status and add mr - test data modifications

Signed-off-by: Isaac Philip <4974658+isaac-philip@users.noreply.github.com>

isaac-philip committed:
feat(functional): #2547 merge-train api for status and add mr - minor changes

Signed-off-by: Isaac Philip <4974658+isaac-philip@users.noreply.github.com>

isaac-philip committed:
feat(functional): #2547 merge-train api for status and add mr - lint formatted by black

Signed-off-by: Isaac Philip <4974658+isaac-philip@users.noreply.github.com>

isaac-philip committed:
feat(functional): #2547 merge-train api for status and add mr - docs added

Signed-off-by: Isaac Philip <4974658+isaac-philip@users.noreply.github.com>

isaac-philip commented:
@nejch   kindly take a look at this PR. 

also if anybody else I should loop in here do share, thanks.

isaac-philip commented:
@JohnVillalovos please let us know for any changes needed before merge, thanks.

JohnVillalovos commented on gitlab/v4/objects/merge_trains.py:
```diff
@@ -1,14 +1,43 @@
+    _update_method: UpdateMethod = UpdateMethod.POST
+
+    _update_attrs = RequiredOptional(
+        optional=("sha", "squash", "when_pipeline_succeeds")
```
Should `auto_merge` be added?

https://docs.gitlab.com/api/merge_trains/#add-a-merge-request-to-a-merge-train

isaac-philip replied:
added

JohnVillalovos commented on tests/unit/objects/test_merge_trains.py:
```diff
@@ -46,6 +46,10 @@
     "duration": 70,
 }
 
+merge_train_update = mr_content.copy()
```
Probably a good idea. Since line 51 would modify the underlying dict.

JohnVillalovos reviewed (changes requested):
Thanks @isaac-philip

Just a few minor changes requested.

Also it could be squashed down to one commit most likely. That can be done when it is merged or you can do it when you update. Either works.

JohnVillalovos committed:
build(release): use correct python-semantic-release/publish-action

The previous 'python-semantic-release/upload-to-gh-release' action has
been deprecated.

Closes: #3210

JohnVillalovos committed:
chore: update to mypy 1.16.0 and resolve issues found

There was an error in the type-hints that was detected when updating
to mypy 1.16.0

The error was:
  $ mypy
  gitlab/v4/cli.py:411: error: Argument 1 to "cls_to_gitlab_resource" has incompatible type "type[Any]"; expected "RESTObject"  [arg-type]
  Found 1 error in 1 file (checked 246 source files)

renovate[bot] committed:
chore(deps): update dependency requests to v2.32.4 [security]

renovate[bot] committed:
chore(deps): update all non-major dependencies

renovate[bot] committed:
chore(deps): update all non-major dependencies

JohnVillalovos committed:
docs: update CONTRIBUTING.rst with policy on issue management

Also update the `stale` job to close issues with no activity, unless
they are assigned to someone.

matthias-reiff committed:
feat(const): add PLANNER_ACCESS constant

JohnVillalovos committed:
ci(stale): improve formatting of stale message

Have it maintain the formatting of the message with the line breaks.

Commit:
chore: release v6.1.0

renovate[bot] committed:
chore(deps): update pre-commit hook maxbrunet/pre-commit-renovate to v41

renovate[bot] committed:
chore(deps): update all non-major dependencies

JohnVillalovos committed:
ci(stale): increase `operations-per-run` to 500

The default value is 30 and it is not able to process all the
issues/PRs. In fact it is failing to make progress. Each day it
processes the same issues/PRs and then stops. It doesn't reach the
other issues/Prs.

Increase to 500, the GitLab limit is actually 3000.

renovate[bot] committed:
chore(deps): update all non-major dependencies

renovate[bot] committed:
chore(deps): update all non-major dependencies

JohnVillalovos committed:
feat: add sync method to force remote mirror updates

Add ProjectRemoteMirror.sync() method to trigger immediate push mirror
updates.

Closes: #3223

Commit:
feat(api): add ListMixin to ProjectMergeRequestDiscussionNoteManager

Closes: #3180

Commit:
feat(api): add ListMixin to ProjectIssueDiscussionNoteManager

renovate[bot] committed:
chore(deps): update all non-major dependencies

renovate[bot] committed:
chore(deps): update dependency furo to v2025

Commit:
chore: release v6.2.0

renovate[bot] committed:
chore(deps): update all non-major dependencies

renovate[bot] committed:
chore(deps): update all non-major dependencies

nickbroon committed:
feat(api): add missing ProjectPackageManager list filters

`package_version` was introduced in GitLab 16.6

See: https://docs.gitlab.com/api/packages/#for-a-project

nickbroon committed:
feat(api): add missing ProjectJob list filters

See: https://docs.gitlab.com/api/jobs/#list-project-jobs

renovate[bot] committed:
chore(deps): update all non-major dependencies

renovate[bot] committed:
chore(deps): update actions/download-artifact action to v5

AdrianDC committed:
feat(users): implement support for 'admins' in administrators 'list'

Signed-off-by: Adrian DC <radian.dc@gmail.com>

renovate[bot] committed:
chore(deps): update all non-major dependencies

renovate[bot] committed:
chore(deps): update actions/checkout action to v5

derekschrock committed:
feat(graphql): update to gql 4.0.0

BREAKING CHANGE: GraphQL.execute() no longer accepts graphql.Source

renovate[bot] committed:
chore(deps): update all non-major dependencies

Commit:
chore: release v6.3.0

renovate[bot] committed:
chore(deps): update actions/setup-python action to v6

renovate[bot] committed:
chore(deps): update all non-major dependencies

AdrianDC committed:
feat(users): implement missing arguments in users 'list'

* feat(users): sort 'user list' arguments against documentation
* feat(users): implement support for 'humans' in users 'list'
* feat(users): implement support for 'public_email' in users 'list'
* feat(users): implement 'created_*/exclude_*/without_*' in users 'list'
* feat(users): implement 'without_projects/skip_ldap/auditors' in administrators 'list'

Signed-off-by: Adrian DC <radian.dc@gmail.com>

AdrianDC committed:
feat(users): implement 'skip_confirmation' in users 'emails' creation

Signed-off-by: Adrian DC <radian.dc@gmail.com>

renovate[bot] committed:
chore(deps): update actions/stale action to v10

renovate[bot] committed:
chore(deps): update all non-major dependencies

renovate[bot] committed:
chore(deps): update all non-major dependencies

Commit:
chore: release v6.4.0

renovate[bot] committed:
chore(deps): update all non-major dependencies

renovate[bot] committed:
chore(deps): update dependency sphinx-autobuild to v2025

renovate[bot] committed:
chore(deps): update all non-major dependencies

JohnVillalovos committed:
docs(pull_mirror): fix incorrect start() method usage example

The documentation incorrectly showed calling start() on the
ProjectPullMirror object (mirror.start()), but the method only
exists on the ProjectPullMirrorManager. Updated the example to
show the correct usage: project.pull_mirror.start().

Fixes #3269

JohnVillalovos committed:
ci(stale): fix permission for stale action and allow manual run

As documented at: https://github.com/actions/stale it requires the
"actions: write" permission.

Also allow manually running the stale action.

JohnVillalovos committed:
fix(semantic-release): enable CHANGELOG.md generation

Also update CHANGELOG.md with missing entries.

Exclude dependency updates from being added to the CHANGELOG.md file in
the future.

Closes: #3209

renovate[bot] committed:
chore(deps): update dependency pytest-cov to v7

JohnVillalovos committed:
feat(api): add content_ref and dry_run_ref parameters to ProjectCiLintManager

Add support for the new GitLab API parameters for validating existing
CI/CD configurations:
- content_ref: specify the branch, tag, or SHA to validate
- dry_run_ref: set branch/tag context for dry run simulations

The deprecated 'ref' parameter is kept for backward compatibility.

Also update documentation with examples showing how to validate
CI configuration from specific branches and with dry run simulation.

Fixes #3260

Commit:
chore: release v6.5.0

renovate[bot] committed:
chore(deps): update dependency pylint to v4

renovate[bot] committed:
chore(deps): update dependency isort to v7

renovate[bot] committed:
chore(deps): update pre-commit hook pycqa/isort to v7

renovate[bot] committed:
chore(deps): update pre-commit hook pycqa/pylint to v4

JohnVillalovos committed:
chore(deps): upgrade coverage to 7.11.0 for Python > 3.9

For Python versions newer than 3.9 use 'coverage==7.11.0'

For 3.9 and older continue to use 'coverage==7.10.7'

renovate[bot] committed:
chore(deps): update github artifact actions

JohnVillalovos committed:
feat!: drop Python 3.9 support and add Python 3.14

Python 3.9 is End-of-Life (EOL) as of 2025-10 as stated in
https://devguide.python.org/versions/ and
https://peps.python.org/pep-0596/#lifespan

By dropping support for Python 3.9 and requiring Python 3.10 or higher
it allows python-gitlab to take advantage of new features in Python
3.10, which are documented at:
https://docs.python.org/3/whatsnew/3.10.html

Also mark Python 3.14 as officially supported as it has been released.

Closes: https://github.com/python-gitlab/python-gitlab/issues/3285

BREAKING CHANGE: As of python-gitlab 7.0.0, Python 3.9 is no longer
supported. Python 3.10 or higher is required.

Commit:
chore: release v7.0.0

JohnVillalovos committed:
ci(release): use the correct token for publish to GitHub

For the: "Publish package distributions to GitHub Releases" use the
correct `secrets.RELEASE_GITHUB_TOKEN`. Before it was using
`secrets.GITHUB_TOKEN`

The other steps are already using `secrets.RELEASE_GITHUB_TOKEN`

Closes: https://github.com/python-gitlab/python-gitlab/issues/3288

renovate[bot] committed:
chore(deps): update all non-major dependencies

renovate[bot] committed:
chore(deps): update all non-major dependencies

renovate[bot] committed:
chore(deps): update all non-major dependencies

renovate[bot] committed:
chore(deps): update all non-major dependencies

renovate[bot] committed:
chore(deps): update dependency pytest to v9

renovate[bot] committed:
chore(deps): update pre-commit hook maxbrunet/pre-commit-renovate to v42

sisp committed:
feat(registry-protection): add support for registry protection rule deletion

renovate[bot] committed:
chore(deps): update all non-major dependencies

renovate[bot] committed:
chore(deps): update actions/checkout action to v6

renovate[bot] committed:
chore(deps): update all non-major dependencies

AlexandreEXFO committed:
fix(utils): prevent negative sleep time in rate limit retry

renovate[bot] committed:
chore(deps): update all non-major dependencies

renovate[bot] committed:
chore(deps): update all non-major dependencies

renovate[bot] committed:
chore(deps): update dessant/lock-threads action to v6

renovate[bot] committed:
chore(deps): update all non-major dependencies

JohnVillalovos committed:
test(functional): update to GitLab 18.6 and resolve issues found

renovate[bot] committed:
chore(deps): update github artifact actions

renovate[bot] committed:
chore(deps): update gitlab/gitlab-ee docker tag to v18.7.0-ee.0 (#3218)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

Commit:
chore: release v7.1.0

renovate[bot] committed:
chore(deps): update all non-major dependencies

renovate[bot] committed:
chore(deps): update all non-major dependencies

renovate[bot] committed:
chore(deps): update gitlab/gitlab-ee docker tag to v18.7.1-ee.0 (#3321)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update all non-major dependencies

isaac-philip committed:
feat(functional): #2547 merge-train api : CR comments changes

isaac-philip commented on docs/gl_objects/merge_trains.rst:
```diff
@@ -9,6 +9,8 @@ Reference
   + :class:`gitlab.v4.objects.ProjectMergeTrain`
   + :class:`gitlab.v4.objects.ProjectMergeTrainManager`
+  + :class:`gilab.v4.objects.ProjectMergeTrainMergeRequest`
+  + :class:`gilab.v4.objects.ProjectMergeTrainMergeRequestManager`
```
made the change
